huggingface_pipelines/speech.py

```diff
@@ -0,0 +1,236 @@
+
+    Example:
+
+        pipeline_config = HFSpeechToEmbeddingPipelineConfig(
+            encoder_model="sonar_speech_encoder",
+            device="cuda",
+            batch_size=16,
+            audio_column="audio"
+        )
+
+        pipeline = HFSpeechToEmbeddingPipeline(pipeline_config)
+    """
+
+    def __init__(self, config: HFSpeechToEmbeddingPipelineConfig):
+        """
+        Initialize the HFSpeechToEmbeddingPipeline.
+
+        Args:
+            config (HFSpeechToEmbeddingPipelineConfig): The configuration for this pipeline.
+        """
+        self.config = config
+        self.model = SpeechToEmbeddingModelPipeline(
+            encoder=self.config.encoder_model,
+            device=torch.device(self.config.device),
+            fbank_dtype=self.config.fbank_dtype
+        )
+
+    def process_batch(self, batch: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Process a batch of audio data, converting it to embeddings.
+
+        This method handles the conversion of audio data to temporary WAV files,
+        generates embeddings using the model, and ensures consistent embedding
+        shapes across the batch.
+
+        Args:
+            batch (Dict[str, Any]): A dictionary containing the batch data.
+                Expected to have an 'audio' key with a list of audio data dictionaries.
+
+        Returns:
+            Dict[str, Any]: The input batch dictionary with an additional key
+                '{audio_column}_embedding' containing the generated embeddings.
+
+        Raises:
+            Exception: If there's an error during batch processing or embedding generation.
+        """
+        try:
+            audio_inputs = []
+            temp_files = []
+
+            for audio_data in batch[self.config.audio_column]:
+                if isinstance(audio_data, dict) and 'array' in audio_data and 'sampling_rate' in audio_data:
+                    temp_file = tempfile.NamedTemporaryFile(
+                        delete=False, suffix='.wav')
+                    temp_files.append(temp_file)
+                    sf.write(temp_file.name,
+                             audio_data['array'], audio_data['sampling_rate'])
+                    audio_inputs.append(temp_file.name)
+                else:
+                    logger.warning(f"Invalid audio data format: {audio_data}")
+
+            if not audio_inputs:
+                logger.warning("No valid audio inputs found in batch.")
+                return batch
+
+            try:
+                all_embeddings = self.model.predict(
+                    input=audio_inputs,
+                    batch_size=self.config.batch_size,
+                    n_parallel=self.config.n_parallel,
+                    pad_idx=self.config.pad_idx
+                )
+
+                # Ensure all embeddings are 2D
+                all_embeddings = [emb.unsqueeze(0) if emb.dim(
```

Comment:
what happens if the audio inputs have multiple channels ?

Reply:
We can convert the multiple channels into mono channels because raising makes the pipeline less durable, we can do this by taking the mean across channels